trace-utils/src/trace_utils.rs

```diff
@@ -136,38 +164,160 @@ pub fn construct_tracer_payload(
     }
 }
 
-pub fn serialize_agent_payload(payload: pb::AgentPayload) -> anyhow::Result<Vec<u8>> {
+pub fn serialize_proto_payload<T>(payload: T) -> anyhow::Result<Vec<u8>>
+where
+    T: ::prost::Message,
+{
     let mut buf = Vec::new();
     buf.reserve(payload.encoded_len());
     payload.encode(&mut buf)?;
     Ok(buf)
 }
 
-pub async fn send(data: Vec<u8>, url: &str, api_key: &str) -> anyhow::Result<()> {
-    let req = Request::builder()
-        .method(Method::POST)
-        .uri(url)
-        .header("Content-type", "application/x-protobuf")
-        .header("DD-API-KEY", api_key)
-        .body(Body::from(data))?;
-
-    let https = HttpsConnectorBuilder::new()
-        .with_native_roots()
-        .https_only()
-        .enable_http1()
-        .build();
-    let client: Client<_, hyper::Body> = Client::builder().build(https);
-    match client.request(req).await {
-        Ok(response) => {
-            if response.status() != StatusCode::ACCEPTED {
-                let body_bytes = hyper::body::to_bytes(response.into_body()).await?;
-                let response_body = String::from_utf8(body_bytes.to_vec()).unwrap_or_default();
-                anyhow::bail!("Server did not accept traces: {response_body}");
+#[derive(Debug, Clone)]
+pub struct SendData {
+    tracer_payloads: Vec<pb::TracerPayload>,
+    size: usize, // have a rough size estimate to force flushing if it's large
+    pub target: Endpoint,
+    headers: HashMap<&'static str, String>,
+}
+
+impl SendData {
+    pub fn new(
+        size: usize,
+        tracer_payload: pb::TracerPayload,
+        tracer_header_tags: TracerHeaderTags,
+        target: &Endpoint,
+    ) -> SendData {
+        let headers = if let Some(api_key) = &target.api_key {
+            HashMap::from([("DD-API-KEY", api_key.as_ref().to_string())])
+        } else {
+            tracer_header_tags.into()
+        };
+
+        SendData {
+            tracer_payloads: vec![tracer_payload],
+            size,
+            target: target.clone(),
+            headers,
+        }
+    }
+
+    pub fn size(&self) -> usize {
+        self.size
+    }
+
+    pub async fn send<'a>(self) -> anyhow::Result<Response<Body>> {
+        let target = &self.target;
+
+        let mut req = hyper::Request::builder()
+            .uri(target.url.clone())
+            .header(hyper::header::USER_AGENT, concat!("Tracer/", env!("CARGO_PKG_VERSION")))
```

Comment:
what's this header used for?

Reply:
Telemetry and profiler use the header to clearly distinguish which code is the sender, in case some things are observed on the receiving side so that it's obvious what code sends it.